torchvision/prototype/datasets/utils/_internal.py

```diff
@@ -250,3 +255,46 @@ def _make_sharded_datapipe(root: str, dataset_size: int) -> IterDataPipe:
     # dp = dp.cycle(2)
     dp = TakerDataPipe(dp, dataset_size)
     return dp
+
+
+def fromfile(
+    file: BinaryIO,
+    *,
+    dtype: torch.dtype,
+    byte_order: str,
+    count: int = -1,
+) -> torch.Tensor:
+    """Construct a tensor from a binary file.
+
+    .. note::
+
+        This function is similar to :func:`numpy.fromfile` with two notable differences:
+
+        1. This function only accepts an open binary file, but not a path to it.
+        2. This function has an additional ``byte_order`` parameter, since PyTorch's ``dtype``'s do not support that
+            concept.
+
+    Args:
+        file (IO): Open binary file.
+        dtype (torch.dtype): Data type of the returned tensor.
+        byte_order (str): Byte order of the data. Can be ``"little"`` or ``"big"`` endian.
+        count (int): Number of values of the returned tensor. If ``-1`` (default), will read the complete file.
+    """
+    byteorder = "<" if byte_order == "little" else ">"
+    char = "f" if dtype.is_floating_point else ("i" if dtype.is_signed else "u")
+    itemsize = (torch.finfo if dtype.is_floating_point else torch.iinfo)(dtype).bits // 8
+    np_dtype = byteorder + char + str(itemsize)
+
+    buffer = file.read(-1 if count == -1 else count * itemsize)
+    # torch.frombuffer can only deal with with the native byte order,
+    # so we use numpy for the I/O and convert to a tensor.
+    return torch.from_numpy(np.frombuffer(buffer, dtype=np_dtype).astype(np_dtype[1:]))
```

Comment:
Also, don't we want to pass `copy=False`?

Reply:
> Could you explain what the `astype(np_dtype[1:])` call does, and why it's needed?

`np.ndarray.astype()` converts the array to the given type. `np_dtype` is a `str` so `np_dtype[1:] == char + str(itemsize)`. The complete call converts the array from an arbitrary endianness to native one, so that it can be read by `torch.frombuffer`. 

> Also, don't we want to pass `copy=False`?

In theory, yes, but in the current implementation, no. `torch.frombuffer` will scream at you if the underlying memory is read-only and without a copy, it is. I'm looking into ways to read from the file and store the data directly in mutable memory.

Reply:
> The complete call converts the array from an arbitrary endianness to native one, so that it can be read by torch.frombuffer.

Thanks, it'd be great to document this

Reply:
A solution to avoid the memcopy is to use a [`mmap`]():

```python
import mmap

import torch

data = torch.rand(5)
path = "data.bin"

data.numpy().tofile(path)

with open(path, "r+b") as file:
    buffer = mmap.mmap(file.fileno(), 0)
    restored_data = torch.frombuffer(buffer, dtype=data.dtype)

assert torch.all(restored_data == data)
```

The only problem with that is the file has to be opened in read / write mode (`"r+"`) and their is currently no interface to do that with datapipes. I'll benchmark first and see if this something that could be helpful.

Reply:
See https://github.com/pytorch/vision/pull/4882#issuecomment-971674863.